File: cfr_deep_stack/algorithms/cfr.py
"""Tabular counterfactual regret minimization.

The implementation follows OpenSpiel's CFR shape: current strategies are
derived through regret matching and the average policy is the convergent output.
It is deliberately compact while the project is still using small poker games
to validate the CFR-average pipeline.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Dict, Mapping, MutableMapping, Sequence

import numpy as np

from cfr_deep_stack.games.kuhn_poker import CHANCE_PLAYER


@dataclass
class InfoStateNode:
    legal_actions: Sequence[int]
    cumulative_regret: MutableMapping[int, float] = field(default_factory=dict)
    cumulative_policy: MutableMapping[int, float] = field(default_factory=dict)


def regret_matching(
    cumulative_regret: Mapping[int, float], legal_actions: Sequence[int]
) -> Dict[int, float]:
    positive_regrets = {
        action: max(float(cumulative_regret.get(action, 0.0)), 0.0)
        for action in legal_actions
    }
    normalizer = sum(positive_regrets.values())
    if normalizer <= 0.0:
        probability = 1.0 / len(legal_actions)
        return {action: probability for action in legal_actions}
    return {
        action: positive_regrets[action] / normalizer for action in legal_actions
    }


class CFRSolver:
    """Alternating-update tabular CFR solver."""

    def __init__(self, game, regret_matching_plus: bool = False):
        self.game = game
        self.num_players = game.num_players()
        self.regret_matching_plus = regret_matching_plus
        self.info_state_nodes: Dict[str, InfoStateNode] = {}
        self.iterations = 0

    def evaluate_and_update_policy(self) -> np.ndarray:
        """Runs one full CFR iteration and returns root values by player."""
        values = np.zeros(self.num_players, dtype=np.float64)
        for player in range(self.num_players):
            reach = np.ones(self.num_players, dtype=np.float64)
            values[player] = self._cfr(self.game.new_initial_state(), player, reach)
        self.iterations += 1
        return values

    def average_policy(self) -> Dict[str, Dict[int, float]]:
        return {
            info_state: self._average_strategy(node)
            for info_state, node in self.info_state_nodes.items()
        }

    def current_policy(self) -> Dict[str, Dict[int, float]]:
        return {
            info_state: regret_matching(node.cumulative_regret, node.legal_actions)
            for info_state, node in self.info_state_nodes.items()
        }

    def action_probabilities(self, state, player: int | None = None) -> Dict[int, float]:
        acting_player = state.current_player() if player is None else player
        info_state = state.information_state_string(acting_player)
        node = self._node_for_state(state, acting_player)
        average = self._average_strategy(node)
        if info_state not in self.info_state_nodes:
            return average
        return average

    def _cfr(self, state, updating_player: int, reach: np.ndarray) -> float:
        if state.is_terminal():
            return float(state.returns()[updating_player])

        current_player = state.current_player()
        if current_player == CHANCE_PLAYER:
            return sum(
                probability * self._cfr(state.child(action), updating_player, reach)
                for action, probability in state.chance_outcomes()
            )

        node = self._node_for_state(state, current_player)
        strategy = regret_matching(node.cumulative_regret, node.legal_actions)

        action_values = {}
        node_value = 0.0
        for action in node.legal_actions:
            next_reach = reach.copy()
            next_reach[current_player] *= strategy[action]
            action_value = self._cfr(state.child(action), updating_player, next_reach)
            action_values[action] = action_value
            node_value += strategy[action] * action_value

        if current_player == updating_player:
            opponent_reach = np.prod(np.delete(reach, updating_player))
            player_reach = reach[updating_player]
            for action in node.legal_actions:
                regret = opponent_reach * (action_values[action] - node_value)
                updated = node.cumulative_regret.get(action, 0.0) + regret
                if self.regret_matching_plus:
                    updated = max(updated, 0.0)
                node.cumulative_regret[action] = updated
                node.cumulative_policy[action] = (
                    node.cumulative_policy.get(action, 0.0)
                    + player_reach * strategy[action]
                )

        return node_value

    def _node_for_state(self, state, player: int) -> InfoStateNode:
        info_state = state.information_state_string(player)
        node = self.info_state_nodes.get(info_state)
        if node is None:
            node = InfoStateNode(tuple(int(action) for action in state.legal_actions(player)))
            self.info_state_nodes[info_state] = node
        return node

    @staticmethod
    def _average_strategy(node: InfoStateNode) -> Dict[int, float]:
        normalizer = sum(node.cumulative_policy.values())
        if normalizer <= 0.0:
            probability = 1.0 / len(node.legal_actions)
            return {action: probability for action in node.legal_actions}
        return {
            action: node.cumulative_policy.get(action, 0.0) / normalizer
            for action in node.legal_actions
        }

File: csrc/src/algorithm/cfr.cc
#include "algorithm/cfr.h"

#include <algorithm>
#include <stdexcept>

namespace fisher::algorithm {
namespace {

std::unordered_map<int, double> RegretMatching(const InfoStateNode& node) {
  std::unordered_map<int, double> strategy;
  double normalizer = 0.0;
  for (int action : node.legal_actions) {
    const auto it = node.cumulative_regret.find(action);
    const double regret = it == node.cumulative_regret.end() ? 0.0 : it->second;
    strategy[action] = std::max(regret, 0.0);
    normalizer += strategy[action];
  }

  if (normalizer <= 0.0) {
    const double probability = 1.0 / static_cast<double>(node.legal_actions.size());
    for (int action : node.legal_actions) strategy[action] = probability;
  } else {
    for (int action : node.legal_actions) strategy[action] /= normalizer;
  }
  return strategy;
}

std::map<int, double> ActionProbabilities(
    const Strategy& strategy, const game::KuhnState& state, int player) {
  const std::string info_state = state.InformationStateString(player);
  const auto it = strategy.find(info_state);
  if (it != strategy.end()) return it->second;

  const auto legal_actions = state.LegalActions();
  const double probability = 1.0 / static_cast<double>(legal_actions.size());
  std::map<int, double> uniform;
  for (int action : legal_actions) uniform[action] = probability;
  return uniform;
}

std::array<double, 2> PolicyValueState(const game::KuhnState& state,
                                       const Strategy& strategy) {
  if (state.IsTerminal()) return state.Returns();

  const int current_player = state.CurrentPlayer();
  std::array<double, 2> values{0.0, 0.0};

  if (current_player == game::kChancePlayer) {
    for (const auto& [action, probability] : state.ChanceOutcomes()) {
      const auto child_values = PolicyValueState(state.Child(action), strategy);
      values[0] += probability * child_values[0];
      values[1] += probability * child_values[1];
    }
    return values;
  }

  for (const auto& [action, probability] :
       ActionProbabilities(strategy, state, current_player)) {
    const auto child_values = PolicyValueState(state.Child(action), strategy);
    values[0] += probability * child_values[0];
    values[1] += probability * child_values[1];
  }
  return values;
}

void CollectInfoStates(const game::KuhnState& state, int player,
                       std::vector<std::string>* out) {
  if (state.IsTerminal()) return;

  const int current_player = state.CurrentPlayer();
  if (current_player == game::kChancePlayer) {
    for (const auto& [action, unused_probability] : state.ChanceOutcomes()) {
      CollectInfoStates(state.Child(action), player, out);
    }
    return;
  }

  if (current_player == player) {
    const std::string info_state = state.InformationStateString(player);
    if (std::find(out->begin(), out->end(), info_state) == out->end()) {
      out->push_back(info_state);
    }
  }

  for (int action : state.LegalActions()) {
    CollectInfoStates(state.Child(action), player, out);
  }
}

double BestResponseValueByEnumeration(const game::KuhnPokerGame& game,
                                      const Strategy& strategy, int player) {
  std::vector<std::string> info_states;
  CollectInfoStates(game.NewInitialState(), player, &info_states);

  double best_value = -1e100;
  const int total_responses = 1 << static_cast<int>(info_states.size());
  for (int mask = 0; mask < total_responses; ++mask) {
    Strategy candidate = strategy;
    for (int i = 0; i < static_cast<int>(info_states.size()); ++i) {
      const int action = (mask >> i) & 1;
      candidate[info_states[i]] = {{action, 1.0}};
    }
    best_value = std::max(
        best_value, PolicyValueState(game.NewInitialState(), candidate)[player]);
  }
  return best_value;
}

}  // namespace

KuhnCfrSolver::KuhnCfrSolver(bool cfr_plus) : cfr_plus_(cfr_plus) {}

void KuhnCfrSolver::Run(const game::KuhnPokerGame& game, int iterations) {
  if (iterations < 0) throw std::invalid_argument("iterations must be non-negative");
  for (int i = 0; i < iterations; ++i) {
    for (int player = 0; player < game.NumPlayers(); ++player) {
      std::array<double, 2> reach{1.0, 1.0};
      Cfr(game.NewInitialState(), player, reach);
    }
    ++iterations_;
  }
}

Strategy KuhnCfrSolver::AveragePolicy() const {
  Strategy policy;
  for (const auto& [info_state, node] : info_state_nodes_) {
    double normalizer = 0.0;
    for (const auto& [unused_action, probability_sum] : node.cumulative_policy) {
      normalizer += probability_sum;
    }

    std::map<int, double> action_probs;
    if (normalizer <= 0.0) {
      const double probability = 1.0 / static_cast<double>(node.legal_actions.size());
      for (int action : node.legal_actions) action_probs[action] = probability;
    } else {
      for (int action : node.legal_actions) {
        const auto it = node.cumulative_policy.find(action);
        action_probs[action] =
            (it == node.cumulative_policy.end() ? 0.0 : it->second) / normalizer;
      }
    }
    policy[info_state] = std::move(action_probs);
  }
  return policy;
}

int KuhnCfrSolver::Iterations() const { return iterations_; }

double KuhnCfrSolver::Cfr(const game::KuhnState& state, int updating_player,
                          const std::array<double, 2>& reach) {
  if (state.IsTerminal()) return state.Returns()[updating_player];

  const int current_player = state.CurrentPlayer();
  if (current_player == game::kChancePlayer) {
    double value = 0.0;
    for (const auto& [action, probability] : state.ChanceOutcomes()) {
      value += probability * Cfr(state.Child(action), updating_player, reach);
    }
    return value;
  }

  InfoStateNode& node = NodeForState(state, current_player);
  const auto strategy = RegretMatching(node);

  std::unordered_map<int, double> action_values;
  double node_value = 0.0;
  for (int action : node.legal_actions) {
    auto next_reach = reach;
    next_reach[current_player] *= strategy.at(action);
    const double action_value = Cfr(state.Child(action), updating_player, next_reach);
    action_values[action] = action_value;
    node_value += strategy.at(action) * action_value;
  }

  if (current_player == updating_player) {
    const double opponent_reach = reach[1 - updating_player];
    const double player_reach = reach[updating_player];
    for (int action : node.legal_actions) {
      const double regret = opponent_reach * (action_values[action] - node_value);
      double updated_regret = node.cumulative_regret[action] + regret;
      if (cfr_plus_) updated_regret = std::max(updated_regret, 0.0);
      node.cumulative_regret[action] = updated_regret;
      node.cumulative_policy[action] += player_reach * strategy.at(action);
    }
  }

  return node_value;
}

InfoStateNode& KuhnCfrSolver::NodeForState(const game::KuhnState& state,
                                           int player) {
  const std::string info_state = state.InformationStateString(player);
  auto [it, inserted] = info_state_nodes_.try_emplace(info_state);
  if (inserted) it->second.legal_actions = state.LegalActions();
  return it->second;
}

std::array<double, 2> PolicyValue(const game::KuhnPokerGame& game,
                                  const Strategy& strategy) {
  return PolicyValueState(game.NewInitialState(), strategy);
}

double NashExploitability(const game::KuhnPokerGame& game,
                          const Strategy& strategy) {
  const auto on_policy = PolicyValue(game, strategy);
  double nash_conv = 0.0;
  for (int player = 0; player < game.NumPlayers(); ++player) {
    nash_conv += BestResponseValueByEnumeration(game, strategy, player) -
                 on_policy[player];
  }
  return nash_conv / static_cast<double>(game.NumPlayers());
}

}  // namespace fisher::algorithm
